armsrc/fpgaloader.c

```diff
@@ -160,9 +164,9 @@ bool FpgaSetupSscDma(uint8_t *buf, int len)
 
 	AT91C_BASE_PDC_SSC->PDC_PTCR = AT91C_PDC_RXTDIS;	// Disable DMA Transfer
 	AT91C_BASE_PDC_SSC->PDC_RPR = (uint32_t) buf;		// transfer to this memory address
-	AT91C_BASE_PDC_SSC->PDC_RCR = len;					// transfer this many bytes
+	AT91C_BASE_PDC_SSC->PDC_RCR = len;					// transfer this many frames
 	AT91C_BASE_PDC_SSC->PDC_RNPR = (uint32_t) buf;		// next transfer to same memory address
-	AT91C_BASE_PDC_SSC->PDC_RNCR = len;					// ... with same number of bytes
+	AT91C_BASE_PDC_SSC->PDC_RNCR = len;					// ... with same number of frames
```

Comment:
No, this is correct. PDC_RNCR contains the number of DMA transfers. The width of each transfer is determined by the peripheral (SSC in this case). 

Simple "proof": If this would be bytes, the DMA transfers would fill only half of the buffer but we would read the whole buffer - which would result in failures.

Reply:
then `sizeof(dmaBuf)` should be DMA_BUFFER_SIZE*2

Reply:
or we just hope no one uses frames > 8 bits?

Reply:
`sizeof(dmaBuf)` **is** DMA_BUFFER_SIZE*2:
```
uint16_t *dmaBuf = (uint16_t*) BigBuf_malloc(ISO14443B_DMA_BUFFER_SIZE * sizeof(uint16_t));
```

Reply:
I just don't like that `iclass.c` and `iso14443a.c` give no hint to the untrained reader, that bufsize is not always equal to dma length. Moreover frame length is hidden in FpgaSetupSsc as a major type argument.

The whole thing is not an implementation error. Just bound to be a pitfall for many.

Reply:
Well, I think that such pitfalls simply cannot be avoided. There is hardware involved and when writing to a peripheral register this has side effects which cannot be understood or even imagined without having a look at the FPGA code and the respective data sheet. The fact that PDC_RNCR counts DMA transfers instead of bytes is just an example. A few comments can help just a bit (and therefore I changed "bytes" - which happened to be the size of the frames up to now - to "frames").

At first I had a separate `FpgaSetSscWidth()` function but then I thought that this would open too much room for faults (like setting to a width which is not supported by the current FPGA mode).

The FPGA generates the SSC frame signal and clock and therefore the FPGA mode determines the transfer width. Therefore I found it most convenient to setup SSC transfers using the FPGA mode as parameter.

If I would be in a position to start writing the ARM from scrach, I would do it more hiearchical. There would be low level functions to access the hardware (like transferring bytes to and from the ARM). On top of these, there would be the demodulators and framing. On top of these then the reader, simulator and snoop functions. Just dreaming a dream ...